spec/data-model.md

```diff
@@ -145,17 +145,22 @@ interface Option {
 }
 ```
 
-A `Reserved` represents an _expression_ with a _reserved_ _annotation_.
-The `sigil` corresponds to the starting sigil of the _reserved_.
+A `Reserved` represents an _expression_ with a _reserved_ or _private-use_ _annotation_.
+The `sigil` corresponds to the starting sigil of the _annotation_.
 The `source` is the "raw" value (i.e. escape sequences are not processed)
-and includes the starting `sigil`.
+and does not include the starting `sigil`.
```

Comment:
Is this true? The escape sequences prevent falsely finding an ending marker. The unescaped `source` is the "raw" value. Reserializing the `source` would reinsert the escapes (this is how one avoids double-escaping traps)

Reply:
The `sigil` in this case is the `^`, `*` or other identified sigil from which we can tell that this annotation is private-use or reserved. Including it or not in `source` is a matter of choice; I'm here proposing that it should not be, as that's how all the other sigils are handled in the spec.

Reply:
My comment here isn't about sigils. It is that preserving the "raw" value without unescaping produces a risk of double-escaping (the receiver cannot tell if the escape sequence is intentional or not). 

What is the value of preserving the "raw" (escaped) representation of the annotation in the data model, given that the purpose of the escapes is either ambiguity in parsing MF2 or in the runtime environment?

Reply:
The problem is that if we process the string, then we need to process it into a sequence of reserved-char sequences and quoted literals, because their escapes are different.

Say we didn't do that, and just applied the escapes. Then we could end up with a cooked source value like
```
one|two|three|four
```

With that, we can't know which of the following would be its original source form:
1. `one|two|three\|four`
2. `one|two\|three|four`
3. `one\|two|three|four`
4. `one\|two\|three\|four`

As the implementation or future spec version undoubtedly cares very much whether a `|` is escaped or not, we would not be able to reconstruct the annotation source.

My presumption is that almost always when encountering unsupported syntax we either don't care about its syntax source, or we just want to pass it through untouched; hence the current approach. If you think this behaviour should be reconsidered, could you raise a separate issue or submit a PR rather than increasing the scope of this one?

Reply:
I think this describes a bug in the syntax, where we were "clever" to smuggle in the quoted production and the whitespace production. I think the argument was something like: "if we unreserved a sigil, then it would just parse correctly". But I think we're better off if reserved and private are opaque to the ABNF. An unreserved sigil is subject to whatever ABNF is applied to it (either the existing annotation syntax or new syntax). Pre-unreserved implementations still won't parse into the sigil's space, so won't be affected.

In other words, I think we should modify the ABNF thusly:

```abnf
private-use    = private-start text
private-start  = "^" / "&"

; reserve additional sigils for use by 
; future versions of this specification
reserved       = reserved-start text
reserved-start = "!" / "@" / "#" / "%" / "*" / "<" / ">" / "/" / "?" / "~"
```

The only escapes in text are around `{` and `}` (and `\` in case one needs `\}` as a character sequence). Brackets retain syntactic meaning everywhere. The current `reserved-body` approach means that implementations parse `reserved` and `private-use` into word tokens and literals, even though the implementation is not allowed to interpret them. I think true opacity is the right approach. (In that case (1), (2), (3), and (4) are different character sequences).

Reply:
(and, yeah, we've had this discussion before [in #374](https://github.com/unicode-org/message-format-wg/pull/374#discussion_r1167447318))

Your argument there was:

> Currently this is a valid expression:
>```
>{:foo key=|{bar}|}
>```
>If we were to unreserve @ and try to give it the same semantics as with :, this would be a parse error due to the unescaped >inner {:
>```
>{@foo key=|{bar}|}
>```

The problem then is: `quoted` allows `{` and `}` unescaped (because they are "quoted"). Adding these characters to the escaped list for `quoted` simplifies the ABNF a tiny amount (quoted-escape and reserve-escape are the same and we lose some productions) and lets implementations parse out expressions by matching (unescaped) `{`/`}` pairs. The cost is that `{`/`}` must be escaped in a literal. 

While fewer escapes is better than more escapes in quoted literals, I don't think that is very onerous compared to having `private-use` (`reserved` scares me less, as we might never use it) be "semi-parsed". I was convinced before that the ABNF would be okay using what we have now because the character sequences could be squeezed in (that I wouldn't actually have to parse the contents). But the data model shows why I was shy in the first place: `private-use` and `reserved` turn out to have parsed structure where we want opacity.

Reply:
Opened #446 to discuss.